### ISSUE TITLE
can't train LM due to constant OOM (on 2 GPUs)

### ISSUE THREAD

**Issue by magdacisowska** [espnet/espnet#2741]
When I try to train LM, no matter how tiny my corpus is, and how tiny the lm config is, I always end up with OOM:
```
RuntimeError: CUDA out of memory. Tried to allocate 9.80 GiB (GPU 0; 39.59 GiB total capacity; 22.09 GiB already allocated; 6.31 GiB free; 31.84 GiB reserved in total by PyTorch)
```
Of course, the more I shrink my model, the later the error occurs, but it occurs anyways.

I run the run.sh script with 2 gpus (NVIDIA A100 with 40 GB memory each)
My lm.yaml:
```
# network architecture
model-module: transformer
att-unit: 32
embed-unit: 64
head: 2
layer: 2
pos-enc: none
unit: 128
accum-grad: 1

# minibatch related
batchsize: 1
maxlen: 5

# optimization related
opt: adam
schedulers: lr=cosine
dropout-rate: 0.0
epoch: 50
gradclip: 1.0
lr: 1e-4
lr-cosine-total: 100000
lr-cosine-warmup: 1000
patience: 0
sortagrad: 0
```

**Comment by sw005320** [comment]
Maybe some of your sentences are too long?
We usually prune the sentences that have more than 300 tokens.
Also, note that our LM training is assuming that each sentence is ended by a line break.

**Comment by magdacisowska** [comment]
I deliberately split my sentences to have max 20 words each, there is no punctuation, all sentences are ended by 'Enter'

**Comment by kan-bayashi** [comment]
How was the encoding?

**Comment by magdacisowska** [comment]
what do you mean?

**Comment by kan-bayashi** [comment]
I mean line break code (LF vs CR+LF).

**Comment by magdacisowska** [comment]
probably CR + LF, and what do you recommend?

**Comment by kan-bayashi** [comment]
Do you use windows?
We assume that unix environment, so CR+LF might cause a problem.
Maybe your training data is loaded as a single sentence, which causes OOM error.

**Comment by magdacisowska** [comment]
yes, I use linux. Okay, maybe thats the problem. I'll check. Thank you

**Comment by magdacisowska** [comment]
Seems like that was not the case. None of the end of line formats work - ``cat -A`` on train/valid files showed ``^M$`` / ``$`` at the end of each line. However, I still get the OOM in both cases.

**Comment by kan-bayashi** [comment]
If it shows `^M$`, the line break code is wrong (Windows' one).

**Comment by magdacisowska** [comment]
Yes, I know, however I checked this possibility as well. As I mentioned before, I work with Linux, so my encoding should be correct from the beginning

**Comment by kan-bayashi** [comment]
Could you show the `train.log`?
It shows something like this:
```
2020-12-03 17:04:25,630 (lm_train:250) INFO: ngpu: 0
2020-12-03 17:04:25,632 (lm_train:253) INFO: python path = (None)
2020-12-03 17:04:25,657 (lm_train:270) INFO: backend = pytorch
2020-12-03 17:04:26,926 (lm:222) INFO: torch version = 1.5.1
2020-12-03 17:04:26,969 (deterministic_utils:26) INFO: torch type check is disabled
2020-12-03 17:04:27,063 (lm_utils:45) INFO: reading text dataset: data/local/lm_train/valid.txt
^M0it [00:00, ?it/s]^M2it [00:00, 3134.76it/s]
2020-12-03 17:04:27,066 (lm_utils:49) INFO: saving binary dataset: data/local/lm_train/valid.txt.h5
2020-12-03 17:04:27,219 (lm_utils:45) INFO: reading text dataset: data/local/lm_train/train.txt
^M0it [00:00, ?it/s]^M3it [00:00, 5061.51it/s]
2020-12-03 17:04:27,221 (lm_utils:49) INFO: saving binary dataset: data/local/lm_train/train.txt.h5
2020-12-03 17:04:27,224 (lm:240) INFO: #vocab = 27
2020-12-03 17:04:27,224 (lm:241) INFO: #sentences in the training data = 3
2020-12-03 17:04:27,225 (lm:242) INFO: #tokens in the training data = 104
2020-12-03 17:04:27,225 (lm:245) INFO: oov rate in the training data = 10.58 %
2020-12-03 17:04:27,225 (lm:247) INFO: #sentences in the validation data = 2
2020-12-03 17:04:27,225 (lm:248) INFO: #tokens in the validation data = 8
```

**Comment by magdacisowska** [comment]
```
# lm_train.py --config conf/lm.yaml --ngpu 2 --backend pytorch --verbose 1 --outdir exp_pl/train_rnnlm_pytorch_lm_bpe40_ngpu2 --tensorboard-dir tensorboard/train_rnnlm_pytorch_lm_bpe40_ngpu2 --train-label data_pl/local/lm_train_bpe40/train.txt --valid-label data_pl/local/lm_train_bpe40/valid.txt --resume --dict data_pl/lang_char/train_960_bpe40_units.txt --dump-hdf5-path data_pl/local/lm_train_bpe40 -e 3 -b 1 --maxlen 5 
# Started at Thu Dec  3 08:10:54 CET 2020
#
2020-12-03 08:10:55,247 (lm_train:250) INFO: ngpu: 2
2020-12-03 08:10:55,247 (lm_train:253) INFO: python path = (None)
2020-12-03 08:10:55,248 (lm_train:270) INFO: backend = pytorch
2020-12-03 08:10:55,732 (lm:222) INFO: torch version = 1.7.0
2020-12-03 08:10:55,732 (deterministic_utils:26) INFO: torch type check is disabled
2020-12-03 08:10:55,790 (lm_utils:40) INFO: loading binary dataset: data_pl/local/lm_train_bpe40/valid.txt.h5
2020-12-03 08:10:55,793 (lm_utils:40) INFO: loading binary dataset: data_pl/local/lm_train_bpe40/train.txt.h5
2020-12-03 08:10:58,248 (lm:240) INFO: #vocab = 63
2020-12-03 08:10:58,248 (lm:241) INFO: #sentences in the training data = 1107624
2020-12-03 08:10:58,249 (lm:242) INFO: #tokens in the training data = 317800333
2020-12-03 08:10:58,249 (lm:243) INFO: oov rate in the training data = 0.04 %
2020-12-03 08:10:58,249 (lm:247) INFO: #sentences in the validation data = 360
2020-12-03 08:10:58,249 (lm:248) INFO: #tokens in the validation data = 110722
2020-12-03 08:10:58,249 (lm:249) INFO: oov rate in the validation data = 0.01 %
2020-12-03 08:10:58,249 (lm:257) INFO: batch size is automatically increased (1 -> 4)
2020-12-03 08:11:01,588 (lm:273) INFO: #iterations per epoch = 553794
2020-12-03 08:11:01,588 (lm:274) INFO: #total iterations = 1661382
2020-12-03 08:11:01,590 (encoder:145) INFO: encoder self-attention layer type = self-attention
2020-12-03 08:11:03,976 (lm:290) INFO: writing a model config file to exp_pl/train_rnnlm_pytorch_lm_bpe40_ngpu2/model.json
epoch       iteration   main/loss   perplexity  val_perplexity  elapsed_time
[J0           100         2.86015     17.3598                     14.2697       
[J     total [..................................................]  0.01%
this epoch [..................................................]  0.02%
       100 iter, 0 epoch / 3 epochs
       inf iters/sec. Estimated time to finish: 0:00:00.
[4A[J0           200         2.71141     14.9231                     27.752        
[J     total [..................................................]  0.01%
this epoch [..................................................]  0.04%
       200 iter, 0 epoch / 3 epochs
    7.4173 iters/sec. Estimated time to finish: 2 days, 14:12:40.357120.
[4A[J0           300         2.67577     14.4307                     41.0738       
[J     total [..................................................]  0.02%
this epoch [..................................................]  0.05%
       300 iter, 0 epoch / 3 epochs
    7.4616 iters/sec. Estimated time to finish: 2 days, 13:50:16.329641.
[4A[J0           400         2.67943     14.4217                     54.7425       
[J     total [..................................................]  0.02%
this epoch [..................................................]  0.07%
       400 iter, 0 epoch / 3 epochs
    7.4124 iters/sec. Estimated time to finish: 2 days, 14:14:40.524997.
[4A[J/home/xxx/third-party-repos/espnet/tools/venv/lib/python3.8/site-packages/torch/nn/parallel/_functions.py:64: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  warnings.warn('Was asked to gather along dimension 0, but all '
Exception in main training loop: Caught RuntimeError in replica 0 on device 0.
Original Traceback (most recent call last):
  File "/home/xxx/third-party-repos/espnet/tools/venv/lib/python3.8/site-packages/torch/nn/parallel/parallel_apply.py", line 61, in _worker
    output = module(*input, **kwargs)
  File "/home/xxx/third-party-repos/espnet/tools/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/xxx/third-party-repos/espnet/espnet/nets/pytorch_backend/lm/transformer.py", line 117, in forward
    h, _ = self.encoder(self.embed(x), self._target_mask(x))
  File "/home/xxx/third-party-repos/espnet/tools/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/xxx/third-party-repos/espnet/espnet/nets/pytorch_backend/transformer/encoder.py", line 294, in forward
    xs, masks = self.encoders(xs, masks)
  File "/home/xxx/third-party-repos/espnet/tools/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/xxx/third-party-repos/espnet/espnet/nets/pytorch_backend/transformer/repeat.py", line 18, in forward
    args = m(*args)
  File "/home/xxx/third-party-repos/espnet/tools/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/xxx/third-party-repos/espnet/espnet/nets/pytorch_backend/transformer/encoder_layer.py", line 81, in forward
    x = residual + self.dropout(self.self_attn(x_q, x, x, mask))
  File "/home/xxx/third-party-repos/espnet/tools/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/xxx/third-party-repos/espnet/espnet/nets/pytorch_backend/transformer/attention.py", line 100, in forward
    scores = torch.matmul(q, k.transpose(-2, -1)) / math.sqrt(self.d_k)
RuntimeError: CUDA out of memory. Tried to allocate 19.50 GiB (GPU 0; 39.59 GiB total capacity; 21.32 GiB already allocated; 16.73 GiB free; 21.42 GiB reserved in total by PyTorch)

# Accounting: time=75 threads=1
# Ended (code 1) at Thu Dec  3 08:12:09 CET 2020, elapsed time 75 seconds

```

**Comment by kan-bayashi** [comment]
Thanks. Then, I have no clear idea...
Maybe one of the sentences in the training data is too long?
Could you check the following command?
```bash
awk < data/local/lm_train/train.txt '{print length($0)}' | sort -n | tail
```

**Comment by magdacisowska** [comment]
I feel like even though I set the --ngpu 2, only one gpu is truly working - gpu0 has load growing up to max memory capacity, while gpu1 has only 2MB load throughout the whole process (until program crashes).
I launched LM training on librispeech for comparison and I see that is that case the GPU load spreads equally to both cards/

The above command results in:
```
2346
2382
2530
2563
2589
2600
2778
3404
3441
3629
```
for my custom dataset, and for librispeech:
```
5778
5861
5938
6253
6360
6427
6433
6853
6890
7754
```

**Comment by magdacisowska** [comment]
Problably the byte pair encoding variant is the problem... for unigram it trains well without errors.
I think that the issue can be deleted as it doesn't introduce any novelty

**Comment by kan-bayashi** [comment]
Then, the reason for this error is caused by your modification?
If so, please fix it by yourself or give us more information about your modification.

**Comment by magdacisowska** [comment]
the only modification I've made is different language and therefore different text corpus. BPE trained well for librispeech, but for my dict and corpus it got OOM. However, when I split the text corpus to have max 80characters in one line, it started to behave properly - no crash, both GPUs are used.
Also, good to remember - it's important to delete .h5 files used for previous training before launching a new one, as its those .h5 files that are actually used as corpora. If they are not deleted, then they are not overwritten with new data. I did a small 'fix' in lm_utils.py, line 39.

So, my errors were:
1. not deleting .h5 files between trials
2. too long lines in txt corpus

Now it trains well
